src/backend/distributed/utils/metadata_cache.c

```diff
@@ -933,6 +934,52 @@ HasUninitializedShardInterval(ShardInterval **sortedShardIntervalArray, int shar
 
 
 /*
+ * HasOverlappingShardInterval determines whether the given list of sorted
+ * shards has overlapping ranges.
+ */
+static bool
+HasOverlappingShardInterval(ShardInterval **shardIntervalArray,
```

Comment:
>  It would be better to enforce this during shard creation.

Right - but I think the users of range partitioning we have in practice actually manipulate shard state themselves. With the current state of out-of-the-box range-partitioning usability, it seems unrealistic to argue they just get to keep the pieces.  If we were able to have a constraint that enforces the boundaries (exclusion constraints would, but we'd have to manipulate catalogs differently), I'd be more welcoming on that front.

 > lower the cost of cache invalidations with high shard counts.

Actually measured that - in comparison to the loading of the shards (way to expensive) and the sorting (also fairly expensive), it doesn't show up in any sort of significant way. In my demo case, on an optimized build, it goes from 1102ms to 1089ms (avg of 3), neutering the check.

I've some fixes that'd make loading faster, but that's for another day/release.

I think it'll be very annoying to debug if we use boundary pruning with overlapping shards for some reason, so I'd rather err on being a bit paranoid.